### PR TITLE
Add Support for joins

### PR DESCRIPTION
**Pull request by bryevdv**
This is still WIP. Wandered down some wrong paths, but I think it is the right general approach.

There is currently some issue with right joins, but other wise this is ready for a look.

### CODE REVIEW

File: modules/cudf/src/table.ts [hunk 1]
```diff
@@ -44,6 +44,41 @@ interface TableConstructor {
...
+  fullJoin(left: Table, right: Table, nullEquality: boolean, memoryResource?: MemoryResource):
+    Column[];
+
+  /**
+   * Returns a pair of row index vectors corresponding to an inner join between the specified
+   * tables.
+   *
+   * @param left_keys The left table
+   * @param right_keys The right table
+   * @param nullEquality controls whether null join-key values
```
Comment (trxcllnt) [inline modules/cudf/src/table.ts:66]:
```suggestion
   * @param nullEquality controls whether null join-key values should match or not
```

File: modules/cudf/src/table.ts [hunk 2]
```diff
@@ -44,6 +44,41 @@ interface TableConstructor {
...
   readCSV<T extends CSVTypeMap = any>(options: ReadCSVOptions<T>):
     {names: (keyof T)[], table: Table};
+
+  /**
+   * Returns a pair of row index vectors corresponding to a full (outer) join between the specified
+   * tables.
+   *
+   * @param left_keys The left table
+   * @param right_keys The right table
+   * @param nullEquality controls whether null join-key values
```
Comment (trxcllnt) [inline modules/cudf/src/table.ts:54]:
```suggestion
   * @param nullEquality controls whether null join-key values should match or not
```

File: modules/cudf/src/data_frame.ts [hunk 3]
```diff
@@ -253,13 +264,75 @@ export class DataFrame<T extends TypeMap = any> {
...
     return new DataFrame(series_map);
   }
 
+  /**
+   * Join columns with other DataFrame.
+   *
+   * @param props the configuration for the join
+   * @returns the joined DataFrame
+   */
+  join<U extends TypeMap, R extends keyof U>(props: JoinProps<U, R>): DataFrame<any> {
```
Comment (bryevdv) [inline modules/cudf/src/data_frame.ts:280]: Definitely welcome any input on if/how the return type here might be improved.

File: modules/cudf/test/dataframe/join-tests.ts [hunk 4]
```diff
@@ -0,0 +1,399 @@
...
+
+  // XXXX This test does not agree with pd/cudf yet
+  // test('can join on multi-index', () => {
+  //   const result = left_double.join({other: right_double, on: ['a', 'b'], how: 'right'});
+  //   expect(result.numColumns).toEqual(4);
+  //   expect(result.names).toEqual(expect.arrayContaining(['a', 'b', 'c', 'd']));
+  //   expect([...result.get('a')]).toEqual([0, 1, 0]);
+  //   expect([...result.get('b')]).toEqual([10, 20, 30]);
+  //   expect([...result.get('c')]).toEqual([1, 2, null]);
+  //   expect([...result.get('d')]).toEqual([10, 30, 20]);
```
Comment (bryevdv) [inline modules/cudf/test/dataframe/join-tests.ts:397]: The values above are pandas results. But the test currently fails
```
   expect(received).toEqual(expected) // deep equality

    - Expected  - 1
    + Received  + 1

      Array [
        0,
        1,
    -   0,
    +   null,
      ]

```

Comment (bryevdv) [reply modules/cudf/test/dataframe/join-tests.ts:397]: I guess the issue is that for right joins it should keep the `"on"` columns from the "right" table. I tried a few things so far but am still searching for the best fix

**Comment by bryevdv** [comment]
Well, this diff fixes the the errant test, but.... ugh. I don't have complete confidence in the rationale for the changes, also its just a mess
```
diff --git a/modules/cudf/src/data_frame.ts b/modules/cudf/src/data_frame.ts
index e0e66c4..1d016ed 100644
--- a/modules/cudf/src/data_frame.ts
+++ b/modules/cudf/src/data_frame.ts
@@ -308,7 +308,11 @@ export class DataFrame<T extends TypeMap = any> {
       }
     }
 
-    right_result = right_result.drop(on) as any;
+    if (how == 'right' && on.length > 1) {
+      left_result = left_result.drop(on as any) as any;
+    } else {
+      right_result = right_result.drop(on) as any;
+    }
 
     // if there are no suffices provided, and there are column conflicts, then we
     // want to drop columns on the right results (so, assign over from left result)
@@ -318,15 +322,29 @@ export class DataFrame<T extends TypeMap = any> {
     const rnames = right_result.names;
 
     // otherwise for any column name conflicts, apply the suffix before combining
-    for (const name of lnames) {
-      if (rnames.includes(name as any)) {
-        const rcol   = right_result.get(name as any);
-        right_result = right_result.drop([name as any]) as any;
-        right_result = right_result.assign({[name + rsuffix]: rcol});
-
-        const lcol  = left_result.get(name as any);
-        left_result = left_result.drop([name] as any) as any;
-        left_result = left_result.assign({[name + lsuffix]: lcol});
+    if (how == 'right') {
+      for (const name of rnames) {
+        if (lnames.includes(name as any)) {
+          const rcol   = right_result.get(name as any);
+          right_result = right_result.drop([name as any]) as any;
+          right_result = right_result.assign({[name + rsuffix]: rcol});
+
+          const lcol  = left_result.get(name as any);
+          left_result = left_result.drop([name] as any) as any;
+          left_result = left_result.assign({[name + lsuffix]: lcol});
+        }
+      }
+    } else {
+      for (const name of lnames) {
+        if (rnames.includes(name as any)) {
+          const rcol   = right_result.get(name as any);
+          right_result = right_result.drop([name as any]) as any;
+          right_result = right_result.assign({[name + rsuffix]: rcol});
+
+          const lcol  = left_result.get(name as any);
+          left_result = left_result.drop([name] as any) as any;
+          left_result = left_result.assign({[name + lsuffix]: lcol});
+        }
       }
     }
```

File: modules/cudf/src/data_frame.ts (hunk 3 above)
Comment (AjayThorve) [reply modules/cudf/src/data_frame.ts:280]: `DataFrame<T&U>`?

Comment (AjayThorve) [reply modules/cudf/src/data_frame.ts:280]: ah understood

File: modules/cudf/src/data_frame.ts [hunk 5]
```diff
@@ -253,13 +264,75 @@ export class DataFrame<T extends TypeMap = any> {
...
+    const {how, other, on} = props;
+    const nullEquality     = props.nullEquality == undefined ? true : props.nullEquality;
+    const lsuffix          = props.lsuffix == undefined ? '' : props.lsuffix;
+    const rsuffix          = props.rsuffix == undefined ? '' : props.rsuffix;
```
Comment (trxcllnt) [inline modules/cudf/src/data_frame.ts:284]: What about this instead?
```suggestion
    const { how, other, on, nullEquality = true, lsuffix = '', rsuffix = '' } = props;
```

File: modules/cudf/src/data_frame.ts [hunk 6]
```diff
@@ -253,13 +264,75 @@ export class DataFrame<T extends TypeMap = any> {
...
+        const rcol   = right_result.get(name);
+        const lcol   = left_result.get(name as any);
+        const merged = lcol.replaceNulls(rcol as any);
+        left_result  = left_result.assign({[name]: merged});
+      }
+    }
+
+    right_result = right_result.drop(on) as any;
+
+    // if there are no suffices provided, and there are column conflicts, then we
```
Comment (trxcllnt) [inline modules/cudf/src/data_frame.ts:313]:
```suggestion
    // if there are no suffixes provided, and there are column conflicts, then we
```

File: modules/cudf/src/data_frame.ts (hunk 5 above)
Comment (bryevdv) [reply modules/cudf/src/data_frame.ts:284]: I keep forgetting this is a thing 😂

File: modules/cudf/src/data_frame.ts (hunk 6 above)
Comment (bryevdv) [reply modules/cudf/src/data_frame.ts:313]: Apropos of nothing, I note that the former head of the graduate math program at UT streadfastly maintained that the correct plural of "octopus" ought to be "octopodes"

**Comment by bryevdv** [comment]
OK I've pushed some changes to get right-joins working but it still seems pretty ugly to me so I will keep looking a bit for improvements

**Comment by bryevdv** [comment]
Also went ahead and added left semi- and anti-joins since there was not alot more to do to include them

**Comment by trxcllnt** [comment]
[Here's an example](https://www.typescriptlang.org/play?#code/KYOwrgtgBAKgngB2FA3gKCpqBVALgMwA4AaDLASRFwGYAmUgXzTQGMAbAQwGcuoARDrg7wkAHhhRgAD1ygAJrxHIAvFA4g4APlRQWAexBdcAJzAtce4wApjwDnINs4UAJZyAXLACUqBlCbs3Lx4RJIy8rwCQkriiMAAdCGE2ii6Bkam5pZWPqlcYEjWSokEhF4A3P7+rJw8UJQ0tGGyIAr8gsJxsUjxDXQpaYYmZhbWuVD5hVbFfbQVVUxouHGwcQCyHAhQqqkA2gDWwHCeGS4gAOYAup5RnUj+5cxnssb4HCzIAMrAxi7AXOJmhF2tE4ikyJhlkhPDBHotnj83h8QRwAGLGDgQYCA6QtNpKDYIcFYKDnYC4UQAaSBrV4hzgenwsE0VnpnkpXk831+-3Eu0pl00cOYUOQACk9GcAArGPQIAEQqAAGQAEp8afj1ptSCSAEpqjWKLUIHVYADyADlDVArKcLlAAGRQemM5UGp0upn6z5eXaXU2YJWfMD4fAuKTWu3nANQXXB0PhyMmM7nNDaHaKgyeS0xtgACy4Nw66Mx2NVn00MeMBaLQhLWNE3srirY+QTUgA-J4gyGw1Iq22+13Y-G+4xHksVhKzpSjqJFVKk78LjGJLjgQTtYrR4n17SJsn7QAfKAAN0lchjlqXKb9pHTUB3Eb3bSjUA7UF2Usu1uvH4ABgAJCgUoMP+UCeIuL50kcroSABwGgcBT5gRBUBASBqGeBhoH-hO+BgCA5guAYUAAFaSiA84kuW1qbiairenRxpXla0E2m+HqwUytFcQyXpqr6-qKj27Y3vaqgAOQAPpSJJMZxr2u7hPub5SdJcCSWmVjSZ404gDKcoAuWxCxmqpk5sqT6mYp7aaLkiqirG2yoIqJJfq4IAHsu5yOs63Fuuq3BQPps5wKIUo2eZVlKf2UCWpo1yBV+lyKn4TroCS7mLmc3kpn5npmUFvChXOkWBTZ1nxRaiWeN6KVpY8JK2LgYDGF54BsGwai8OozjBbc9bYigHm5YVupJbqKUMEKaCLPoQxQLY+RsLg0kAIwuRRZxWJlWBZp+kn4HoeiScJJKtrFngyfmXDyYqxiDuG13SdWd25jWUCdd1wV9T1KJDaIqTHXonizKZABGHDGJ4ST+M2zWfd9-1-QNxYYg2wMnWDVB0JD0Ow6U+MAF44408OMBUzALUYS3-GAq3SU0qjbSAu2ZiAni7EdJ1nR9hZfQzP29Ro-2DRjw1QCDZN0BTD1I0LKOi2jdYS0DUvY-UuP0FAUOk1r5MzZTE407gdMrWt1BbZR7Mkgd3Mg3zLZPVIL23fdzUuy9b0e1gt2eMjv3K5E6Olur0sG3jusEzgpRy4jAuByL-Uh6rYdY6Dkc61DMOx0QJMy00RtzVTrDpGby0M2tAAs1s7XtmD2zzp2mZJOdOxdCtdUrKcA2rGeF-judw8XCcB4rQe9+L6ca5n4PR8PRPR-rsxywwpem+bVfSQArHXbMN1ATeO637fnX7XfC2owd9zPEfzznhP58vhfx1gb3j93k9i6HmOz4PC8n4kBflnNeVMgA) of one way to do the `join` generic type signature in the TS playground:

image

File: modules/cudf/src/table.ts [hunk 7]
```diff
@@ -44,6 +44,63 @@ interface TableConstructor {
...
+   * Returns a pair of row index vectors corresponding to a full (outer) join between the specified
+   * tables.
+   *
+   * @param left_keys The left table
+   * @param right_keys The right table
+   * @param nullEquality controls whether null join-key values should match or not
+   * @param memoryResource An optional MemoryResource used to allocate the result's device memory.
+   */
+  fullJoin(left: Table, right: Table, nullEquality: boolean, memoryResource?: MemoryResource):
+    Column[];
```
Comment (trxcllnt) [inline modules/cudf/src/table.ts:58]:
```suggestion
    [Column<Int32>, Column<Int32>];
```

File: modules/cudf/src/table.ts [hunk 8]
```diff
@@ -44,6 +44,63 @@ interface TableConstructor {
...
+   * Returns a pair of row index vectors corresponding to an inner join between the specified
+   * tables.
+   *
+   * @param left_keys The left table
+   * @param right_keys The right table
+   * @param nullEquality controls whether null join-key values should match or not
+   * @param memoryResource An optional MemoryResource used to allocate the result's device memory.
+   */
+  innerJoin(left: Table, right: Table, nullEquality: boolean, memoryResource?: MemoryResource):
+    Column[];
```
Comment (trxcllnt) [inline modules/cudf/src/table.ts:70]:
```suggestion
    [Column<Int32>, Column<Int32>];
```

File: modules/cudf/src/table.ts [hunk 9]
```diff
@@ -44,6 +44,63 @@ interface TableConstructor {
...
+  /**
+   * Returns a pair of row index vectors corresponding to a left join between the specified tables.
+   *
+   * @param left_keys The left table
+   * @param right_keys The right table
+   * @param nullEquality controls whether null join-key values should match or not
+   * @param memoryResource An optional MemoryResource used to allocate the result's device memory.
+   */
+  leftJoin(left: Table, right: Table, nullEquality: boolean, memoryResource?: MemoryResource):
+    Column[];
```
Comment (trxcllnt) [inline modules/cudf/src/table.ts:81]:
```suggestion
    [Column<Int32>, Column<Int32>];
```

File: modules/cudf/src/table.ts [hunk 10]
```diff
@@ -44,6 +44,63 @@ interface TableConstructor {
...
+  /**
+   * Returns an index vectors corresponding to a left semijoin between the specified tables.
+   *
+   * @param left_keys The left table
+   * @param right_keys The right table
+   * @param nullEquality controls whether null join-key values should match or not
+   * @param memoryResource An optional MemoryResource used to allocate the result's device memory.
+   */
+  leftSemiJoin(left: Table, right: Table, nullEquality: boolean, memoryResource?: MemoryResource):
+    Column;
```
Comment (trxcllnt) [inline modules/cudf/src/table.ts:92]:
```suggestion
    Column<Int32>;
```

File: modules/cudf/src/table.ts [hunk 11]
```diff
@@ -44,6 +44,63 @@ interface TableConstructor {
...
+  /**
+   * Returns an index vectors corresponding to a left antijoin between the specified tables.
+   *
+   * @param left_keys The left table
+   * @param right_keys The right table
+   * @param nullEquality controls whether null join-key values should match or not
+   * @param memoryResource An optional MemoryResource used to allocate the result's device memory.
+   */
+  leftAntiJoin(left: Table, right: Table, nullEquality: boolean, memoryResource?: MemoryResource):
+    Column;
```
Comment (trxcllnt) [inline modules/cudf/src/table.ts:103]:
```suggestion
    Column<Int32>;
```

**Comment by trxcllnt** [comment]
Pushed commits with the generic signatures that include `lsuffix` and `rsuffix` :tada: 

image

**Pull request merged**